politeiad/backendv2/tstorebe/plugins/pi/cmds_test.go

```diff
@@ -0,0 +1,262 @@
+)
+
+// billingStatusTest contains the input and output for a test that tests
+// the cmdBillingStatus func.
+type billingStatusTest struct {
+	name  string    // Test name
+	input testInput // Input
+	err   error     // Expected output
+}
+
+type testInput struct {
+	token   []byte
+	payload string
+}
+
+func TestCmdBillingStatus(t *testing.T) {
+	// Setup pi plugin
+	p, cleanup := newTestPiPlugin(t)
+	defer cleanup()
+
+	// Setup an identity that will be used to create the payload
+	// signatures.
+	fid, err := identity.New()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Setup test data
+	var (
+		// Valid input
+		token     = "45154fb45664714b"
+		status    = pi.BillingStatusCompleted
+		publicKey = fid.Public.String()
+
+		msg        = token + strconv.Itoa(int(status))
+		signatureb = fid.SignMessage([]byte(msg))
+		signature  = hex.EncodeToString(signatureb[:])
+
+		// signatureIsWrong is a valid hex encoded, ed25519 signature,
+		// but that does not correspond to the valid input parameters
+		// listed above.
+		signatureIsWrong = "b387f678e1236ca1784c4bc77912c754c6b122dd8b" +
+			"3e499617706dd0bd09167a113e59339d2ce4b3570af37a092ba88f39e7f" +
+			"c93a5ac7513e52dca3e5e13f705"
+	)
+	tokenb, err := hex.DecodeString(token)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Setup tests
+	var tests = []struct {
+		name  string // Test name
+		token []byte
+		sbs   pi.SetBillingStatus
+		err   error // Expected error output
+	}{
+		{
+			"payload token invalid",
+			tokenb,
+			setBillingStatus(t, fid,
+				pi.SetBillingStatus{
+					Token:  "zzz",
+					Status: status,
+					Reason: "",
+				}),
+			pluginError(pi.ErrorCodeTokenInvalid),
+		},
+		{
+			"payload token does not match cmd token",
+			tokenb,
+			setBillingStatus(t, fid,
+				pi.SetBillingStatus{
+					Token:  "da70d0766348340c",
+					Status: status,
+					Reason: "",
+				}),
+			pluginError(pi.ErrorCodeTokenInvalid),
+		},
+		{
+			"signature is not hex",
+			tokenb,
+			pi.SetBillingStatus{
+				Token:     token,
+				Status:    status,
+				Reason:    "",
+				PublicKey: publicKey,
+				Signature: "zzz",
+			},
+			pluginError(pi.ErrorCodeSignatureInvalid),
+		},
+		{
+			"signature is the wrong size",
+			tokenb,
+			pi.SetBillingStatus{
+				Token:     token,
+				Status:    status,
+				Reason:    "",
+				PublicKey: publicKey,
+				Signature: "123456",
+			},
+			pluginError(pi.ErrorCodeSignatureInvalid),
+		},
+		{
+			"signature is wrong",
+			tokenb,
+			pi.SetBillingStatus{
+				Token:     token,
+				Status:    status,
+				Reason:    "",
+				PublicKey: publicKey,
+				Signature: signatureIsWrong,
+			},
+			pluginError(pi.ErrorCodeSignatureInvalid),
+		},
+		{
+			"public key is not a hex",
+			tokenb,
+			pi.SetBillingStatus{
+				Token:     token,
+				Status:    status,
+				Reason:    "",
+				PublicKey: "",
+				Signature: signature,
+			},
+			pluginError(pi.ErrorCodePublicKeyInvalid),
+		},
+		{
+			"set billing status to active",
```

Comment:
Can you order these tests in the same order that they occur in the code.  It makes it easy to check that all error paths have been tested when you can just read through the implemeantion code and check the error path against the next test case.

Reply:
💯